### PR TITLE
Adding AD version of MatchedValueBC

### LINKED ISSUE #18212: Automatic differentiation version of MatchedValueBC
## Reason

Add `ADMatchedValueBC` to support the task of adding automatic-differentiation to THM.

## Design

Use `MatchedValueBC` and convert it into AD-version.

## Impact

Gain capability, support development of THM.

### PR DESCRIPTION
**Pull request by andrsd**
Closes #18212

### CODE REVIEW

File: framework/include/bcs/ADMatchedValueBC.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADNodalBC.h"
+
+class ADMatchedValueBC;
+
+template <>
+InputParameters validParams<ADMatchedValueBC>();
```
Comment (lindsayad) [inline framework/include/bcs/ADMatchedValueBC.h:17]: Can remove

File: framework/include/bcs/ADMatchedValueBC.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  ADMatchedValueBC(const InputParameters & parameters);
+
+protected:
+  virtual ADReal computeQpResidual() override;
+
+  const VariableValue & _v;
```
Comment (lindsayad) [inline framework/include/bcs/ADMatchedValueBC.h:32]:
```suggestion
  const ADVariableValue & _v;
```

File: framework/src/bcs/ADMatchedValueBC.C [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADMatchedValueBC.h"
+
+registerMooseObject("MooseApp", ADMatchedValueBC);
+
+defineLegacyParams(ADMatchedValueBC);
+
```
Comment (lindsayad) [inline framework/src/bcs/ADMatchedValueBC.C:15]: Can remove

File: framework/src/bcs/ADMatchedValueBC.C [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+{
+  InputParameters params = ADNodalBC::validParams();
+  params.addRequiredCoupledVar("v", "The variable whose value we are to match.");
+  params.addClassDescription("Implements a nodal BC which equates two different Variables' values "
+                             "on a specified boundary.");
+  return params;
+}
+
+ADMatchedValueBC::ADMatchedValueBC(const InputParameters & parameters)
+  : ADNodalBC(parameters), _v(coupledValue("v"))
```
Comment (lindsayad) [inline framework/src/bcs/ADMatchedValueBC.C:27]:
```suggestion
  : ADNodalBC(parameters), _v(adCoupledValue("v"))
```

File: test/tests/bcs/ad_matched_value_bc/tests [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[Tests]
```
Comment (lindsayad) [inline test/tests/bcs/ad_matched_value_bc/tests:1]: As for the other PR you may want to add a `PetscJacobianTest`

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: test/tests/bcs/ad_matched_value_bc/tests (hunk 5 above)
Comment (andrsd) [reply test/tests/bcs/ad_matched_value_bc/tests:1]: 👍

File: framework/include/bcs/ADMatchedValueBC.h (hunk 1 above)
Comment (andrsd) [reply framework/include/bcs/ADMatchedValueBC.h:17]: Ok. I found this in the original file, so I thought we were still doing that.

**Review by lindsayad — APPROVED** [review]

**Comment by andrsd** [comment]
Thanks for the review!

**Pull request merged**